feature/text_based.py

```diff
@@ -0,0 +1,694 @@
+            # Select a sample of selection_size without repetition
+            selected = []
+            while len(selected) < selected_size and self._num_cols is not None:
+                i = random.randint(0, self._num_cols - 1)
+                while i in selected:
+                    i = random.randint(0, self._num_cols -1)
+                selected.append(i)
+
+            # Count covered categories
+            num_row_covered = self._get_num_row_covered(selected)
+            if num_row_covered > best_covered:
+                best_covered = num_row_covered
+                best_selected = selected
+
+        # Calculate coverage metrics
+        num_row_covered = self._get_num_row_covered(best_selected)
+        coverage = num_row_covered / self._num_rows
+
+        if self.verbose:
+            print("\nRANDOM SELECTION:", selected_size, "columns to cover rows ", self._num_rows)
+            print("=" * 40)
+            print("SIZE:", len(best_selected),
+                  "reduction: {:.2f}".format((self._num_cols - len(best_selected)) / self._num_cols))
+            print("SELECTED:", best_selected)
+            print("NUM (AVG) ROWS COVERED:",
+                  num_row_covered, "coverage: {:.2f}".format(coverage))
+            print("STATUS: RANDOM")
+            print("Cost Metric:", self.cost_metric)
+            print("=" * 40)
+        return best_selected
+
+    def _solve_set_cover(self, data: Data) -> List:
+
+        # Create Model object
+        model = Model("Set Cover Model")
+
+        # Variables
+        x = [model.add_var(var_type=BINARY) for _ in data.X]
+
+        # Constraint: every row should be covered
+        for row in data.rows:
+            model.add_constr(xsum(data.matrix[row, i] * x[i] for i in data.X) >= 1)
+
+        # Objective: minimize
+        model.objective = minimize(xsum(data.cost[i] * x[i] for i in data.X))
+
+        # Solve (optimize using Gurobi)
+        model.verbose = 0
+        model.optimize()
+        check_true(model.status == OptimizationStatus.OPTIMAL, "Max Cover Error: optimal solution not found.")
+
+        # Solution
+        selected = [i for i in data.X if float(x[i].x) >= 0.99]
+
+        if self.verbose:
+            print("=" * 40)
+            print("SET COVER OBJECTIVE:", model.objective_value)
+            print("SELECTED:", selected)
+            print("STATUS:", model.status)
+            print("=" * 40)
+
+        # Return
+        return selected
+
+    def _solve_set_cover_cost_option(self, cost: np.ndarray) -> List:
+        data = Data(cost=cost, matrix=self.matrix)
+        selected = self._solve_set_cover(data)
+        return selected
+
+    def _get_diversity_cost(self, selected_size: int) -> List[float]:
+        """
+        diversity cost:
+            there is a situation where selected_size is equal to the number of rows in features and all diversity
+             costs become zero. There are two ways to handle this situation:
```

Comment:
can you add a comment on why the code picks or the other

Reply:
The reason why we need to add NEXT_S_DIS or DUMMY is that there are two alternative options to handle situations where the minimum distance between each two features is zero. This situation may occur if the features are very similar or if there are a limited number of features. When this happens, the diversity cost becomes zero. Therefore, the code adds NEXT_S_DIS to replace the zero value in the distances matrix with the next smallest non-zero distance. This ensures that each feature has a non-zero diversity cost, which helps to select a more diverse set of features even when the minimum distance between each two features is zero. Also, DUMMY may not be a good option (it is @takojunior suggestion) for Greedy since the adjusted cost brings all values between (0, 1). In this case, there is not much difference between uni-cost and diverse cost, since the adjusted cost elements are so similar. We need to decide with which option we want to continue.
I also copied this comment for the same question @bkleyn asked.